core/src/main/java/org/elasticsearch/action/search/SearchTransportService.java

```diff
@@ -76,10 +76,13 @@
     public static final String FETCH_ID_ACTION_NAME = "indices:data/read/search[phase/fetch/id]";
 
     private final TransportService transportService;
+    private final ResponseCollectorService responseCollectorService;
 
-    public SearchTransportService(Settings settings, TransportService transportService) {
+    public SearchTransportService(Settings settings, TransportService transportService,
+                                  ResponseCollectorService responseCollectorService) {
```

Comment:
I played with this, however, it requires making `SearchExecutionStatsCollector` public instead of package private, since we need to construct it within the `BiFunction` passed in to `SearchTransportService` in `Node`.

I think this is what you had in mind, but let me know if I'm mistaken:

```diff
diff --git a/core/src/main/java/org/elasticsearch/action/search/SearchExecutionStatsCollector.java b/core/src/main/java/org/elasticsearch/action/search/SearchExecutionStatsCollector.java                                                   
index 031e860..96d5e18 100644
--- a/core/src/main/java/org/elasticsearch/action/search/SearchExecutionStatsCollector.java
+++ b/core/src/main/java/org/elasticsearch/action/search/SearchExecutionStatsCollector.java
@@ -23,8 +23,10 @@ import org.elasticsearch.action.ActionListener;
 import org.elasticsearch.node.ResponseCollectorService;
 import org.elasticsearch.search.SearchPhaseResult;
 import org.elasticsearch.search.query.QuerySearchResult;
+import org.elasticsearch.transport.Transport;
 
 import java.util.Objects;
+import java.util.function.BiFunction;
 
 /**
  * A wrapper of search action listeners (search results) that unwraps the query
@@ -47,6 +49,10 @@ final class SearchExecutionStatsCollector implements ActionListener<SearchPhaseR
         this.nodeId = nodeId;
     }
 
+    public static BiFunction<Transport.Connection, SearchActionListener, ActionListener> makeWrapper(ResponseCollectorService service) {                                                                                                    
+        return (connection, originalListener) -> new SearchExecutionStatsCollector(listener, service, connection.getNode().getId());                                                                                                        
+    }
+
     @Override
     public void onResponse(SearchPhaseResult response) {
         QuerySearchResult queryResult = response.queryResult();
diff --git a/core/src/main/java/org/elasticsearch/action/search/SearchTransportService.java b/core/src/main/java/org/elasticsearch/action/search/SearchTransportService.java                                                                 
index 2587623..3ad1517 100644
--- a/core/src/main/java/org/elasticsearch/action/search/SearchTransportService.java
+++ b/core/src/main/java/org/elasticsearch/action/search/SearchTransportService.java
@@ -56,6 +56,7 @@ import org.elasticsearch.transport.TransportResponse;
 import org.elasticsearch.transport.TransportService;
 
 import java.io.IOException;
+import java.util.function.BiFunction;
 import java.util.function.Supplier;
 
 /**
@@ -77,12 +78,13 @@ public class SearchTransportService extends AbstractComponent {
 
     private final TransportService transportService;
     private final ResponseCollectorService responseCollectorService;
+    private final BiFunction<Transport.Connection, SearchActionListener, ActionListener> responseWrapper;
 
     public SearchTransportService(Settings settings, TransportService transportService,
-                                  ResponseCollectorService responseCollectorService) {
+                                  BiFunction<Transport.Connection, SearchActionListener, ActionListener> responseWrapper) {                                                                                                                 
         super(settings);
         this.transportService = transportService;
-        this.responseCollectorService = responseCollectorService;
+        this.responseWrapper = responseWrapper;
     }
 
     public void sendFreeContext(Transport.Connection connection, final long contextId, OriginalIndices originalIndices) {                                                                                                                   
@@ -123,12 +125,7 @@ public class SearchTransportService extends AbstractComponent {
         final boolean fetchDocuments = request.numberOfShards() == 1;
         Supplier<SearchPhaseResult> supplier = fetchDocuments ? QueryFetchSearchResult::new : QuerySearchResult::new;
 
-        final ActionListener handler;
-        if (responseCollectorService == null) {
-            handler = listener;
-        } else {
-            handler = new SearchExecutionStatsCollector(listener, responseCollectorService, connection.getNode().getId());                                                                                                                  
-        }
+        final ActionListener handler = responseWrapper.apply(connection, listener);
         transportService.sendChildRequest(connection, QUERY_ACTION_NAME, request, task,
             new ActionListenerResponseHandler<>(handler, supplier));
     }
diff --git a/core/src/main/java/org/elasticsearch/node/Node.java b/core/src/main/java/org/elasticsearch/node/Node.java
index bb2264f..b10f645 100644
--- a/core/src/main/java/org/elasticsearch/node/Node.java
+++ b/core/src/main/java/org/elasticsearch/node/Node.java
@@ -29,6 +29,7 @@ import org.elasticsearch.ElasticsearchTimeoutException;
 import org.elasticsearch.Version;
 import org.elasticsearch.action.ActionModule;
 import org.elasticsearch.action.GenericAction;
+import org.elasticsearch.action.search.SearchExecutionStatsCollector;
 import org.elasticsearch.action.search.SearchPhaseController;
 import org.elasticsearch.action.search.SearchTransportService;
 import org.elasticsearch.action.support.TransportAction;
@@ -424,8 +425,8 @@ public class Node implements Closeable {
             final TransportService transportService = newTransportService(settings, transport, threadPool,
                 networkModule.getTransportInterceptor(), localNodeFactory, settingsModule.getClusterSettings());
             final ResponseCollectorService responseCollectorService = new ResponseCollectorService(this.settings, clusterService);                                                                                                          
-            final SearchTransportService searchTransportService =  new SearchTransportService(settings,
-                transportService, responseCollectorService);
+            final SearchTransportService searchTransportService =  new SearchTransportService(settings, transportService,                                                                                                                   
+                    SearchExecutionStatsCollector.makeWrapper(responseCollectorService));
             final Consumer<Binder> httpBind;
             final HttpServerTransport httpServerTransport;
             if (networkModule.isHttpEnabled()) {
```